src/backend/joanie/core/factories.py

```diff
@@ -221,70 +229,160 @@ def resource_link(self):
         )
 
     @factory.lazy_attribute
-    def start(self):  # pylint: disable=no-self-use
+    def start(self):
         """
-        A start datetime for the course run is chosen randomly in the short past/future (it can
-        of course be forced if we want something else), then the other significant dates
-        for the course run are chosen randomly in periods that make sense with this start date.
+        Compute a start date according to the course run state and the ref date. Otherwise, pick
+        a random date in the past/future in the range of 1 to 365 days.
         """
-        now = django_timezone.now()
-        period = timedelta(days=200)
+        if self.state == CourseState.TO_BE_SCHEDULED:
+            return None
+
+        period = timedelta(
+            days=random.randrange(1, 365, 1)  # nosec
+        )  # between 1 and 365 days
+
+        if self.state in [
+            CourseState.ONGOING_OPEN,
+            CourseState.ONGOING_CLOSED,
+            CourseState.ARCHIVED_OPEN,
+            CourseState.ARCHIVED_CLOSED,
+        ]:
+            # The course run is on going or archived,
+            # so the start date must be less than the ref date
+            min_date = self.ref_date - period
+            max_date = self.ref_date
+        elif self.state in [
+            CourseState.FUTURE_OPEN,
+            CourseState.FUTURE_NOT_YET_OPEN,
+            CourseState.FUTURE_CLOSED,
+        ]:
+            # The course run has not yet started,
+            # so the start date must be greater than the ref date
+            min_date = self.ref_date
+            max_date = self.ref_date + period
+        else:
+            # Otherwise, the start date can be chosen randomly in the past/future
+            min_date = self.ref_date - period
+            max_date = self.ref_date + period
+
         return datetime.utcfromtimestamp(
             random.randrange(  # nosec
-                int((now - period).timestamp()), int((now + period).timestamp())
+                int(min_date.timestamp()), int(max_date.timestamp())
             )
         ).replace(tzinfo=timezone.utc)
 
     @factory.lazy_attribute
     def end(self):
         """
-        The end datetime is at a random duration after the start datetme (we pick within 90 days).
+        Compute the end date according to the course run state and the ref date.
+        Otherwise, pick a random date in the range of 1 to 365 days after the start date
         """
         if not self.start:
             return None
-        period = timedelta(days=90)
+
+        period = timedelta(
+            days=random.randrange(1, 365, 1)  # nosec
+        )  # between 1 and 365 days
+
+        if self.state in [CourseState.ARCHIVED_OPEN, CourseState.ARCHIVED_CLOSED]:
+            # The course run is archived, end date must be less than ref date
+            min_date = self.start
+            max_date = self.ref_date
```

Comment:
we are not sure that self.start is less than self.ref_date are we? It could happen when `start` and/or `ref_date` are forced and in this case we should raise an error when things are not coherent?